apps/api/src/domain/models/term/tokenization/ChilcotinAlphabetParser.ts

```diff
@@ -0,0 +1,259 @@
+// import { isNonEmptyObject } from '@coscrad/validation-constraints';
+// import { InternalError } from 'apps/api/src/lib/errors/InternalError';
+
+import assert = require('node:assert');
+import { InternalError } from '../../../../lib/errors/InternalError';
+import { AlphabetCharacters } from './tokenizer.interface';
+
+class Node {
+    text: string;
+
+    transitions: Map<string, Node> = new Map();
+
+    isLetter: boolean;
+
+    constructor(text: string, isLetter = true) {
+        this.text = text;
+
+        this.isLetter = isLetter;
+    }
+
+    registerTransition(keystroke: string, newNode: Node): this {
+        if (this.transitions.has(keystroke)) {
+            throw new InternalError(
+                `keystroke: ${keystroke} cannot cause ${this.text} to transition to ${
+                    newNode.text
+                }, as it already transtions to: ${this.transitions.get(keystroke).text}`
+            );
+        }
+
+        this.transitions.set(keystroke, newNode);
+
+        return this;
+    }
+
+    size(): number {
+        return (
+            this.transitions.size +
+            [...this.transitions.values()].reduce((count, nextNode) => count + nextNode.size(), 0)
+        );
+    }
+
+    toList(list?: string[]): string[] {
+        if (!list) {
+            list = [];
+        }
+
+        if (this.text !== null && this.isLetter) {
+            list.push(this.text);
+        }
+
+        for (const childNode of this.transitions.values()) {
+            childNode.toList(list);
+        }
+
+        return list;
+    }
+
+    hasTransition(keystroke: string): boolean {
+        return this.transitions.has(keystroke);
+    }
+
+    transition(keystroke: string): Node {
+        if (!this.hasTransition(keystroke)) {
+            throw new InternalError(
+                `No transition found for keystroke: ${keystroke}. Did you forget to check "hasTransition(${keystroke})?" `
+            );
+        }
+
+        return this.transitions.get(keystroke);
+    }
+}
+
+const isolatedLetters = 'aeiɨuobpmnjzŝẑŵhyʔ'.split('');
+
+export class ChilcotinAlphabetParser {
+    // we are flagging that the root is not a letter
+    private root: Node = new Node(null, false);
+
+    private punctuation: Set<string>;
+
+    constructor(punctuationNativeCharacterss: string[] = `"?.!-,`.split('')) {
```

Comment:
Would this be reuseable for other languages \ alphabets? If so, we would need to register this against a language code and make this available via site configuration at some point.